code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.kt

```diff
@@ -0,0 +1,279 @@
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.optimize
+
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.util.DataReader
+import com.adobe.marketing.mobile.util.DataReaderException
+import org.json.JSONObject
+import java.lang.ref.SoftReference
+import java.util.Objects
+
+class Offer private constructor() {
+
+    companion object {
+        private const val SELF_TAG = "Offer"
+
+        @JvmStatic
+        fun fromEventData(data: Map<String, Any?>?): Offer? {
+            if (OptimizeUtils.isNullOrEmpty(data)) {
+                Log.debug(
+                    OptimizeConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Cannot create Offer object, provided data Map is empty or null."
+                )
+                return null
+            }
+            return try {
+                val id = DataReader.getString(data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ID)
+                val etag = DataReader.getString(data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_ETAG)
+                val score = DataReader.optDouble(data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCORE, 0.0)
+                val schema = DataReader.getString(data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_SCHEMA)
+                val meta = DataReader.getTypedMap(Any::class.java, data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_META)
+                val offerData = DataReader.getTypedMap(Any::class.java, data, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA)
+
+                if (!OptimizeUtils.isNullOrEmpty(offerData)) {
+                    val nestedId = DataReader.getString(offerData, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_ID)
+                    if (nestedId.isNullOrEmpty() || nestedId != id) {
+                        Log.debug(
+                            OptimizeConstants.LOG_TAG,
+                            SELF_TAG,
+                            "Cannot create Offer object, provided item id is null or empty or doesn't match item data id."
+                        )
+                        return null
+                    }
+                    val format = DataReader.getString(offerData, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_FORMAT)
+                    val offerType = format?.let { OfferType.from(it) }
+                        ?: OfferType.from(DataReader.getString(offerData, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_TYPE))
+                    val language = DataReader.getStringList(offerData, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_LANGUAGE)
+                    val characteristics = DataReader.getStringMap(offerData, OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CHARACTERISTICS)
+
+                    val content = getContentFromOfferData(offerData) ?: DataReader.optString(
```

Comment:
`getContentFromOfferData` never returns an empty string. The `?:` block is never executed